add support for backing up modules with --module-only via --backup-modules

damianam opened the pull request:
With this PR easybuild will create a backup of existing modules when using `--modules-only`. 2 extra comments:

1. While I was at it, I also used the new `back_up_file` function to rotate logs
2. I think it makes sense to print the `diff` of modules in `stdout`. Thoughts?

boegel commented:
@damianam This needs more work (e.g. review & tests), and since neither of us will have time to work on this in the coming days, let's bump the milestone to `3.3.0` on this and revisit it later.

damianam commented:
@boegel take a look when you can. I'll see if tomorrow I can write some tests for it.

boegel commented on easybuild/framework/easyblock.py:
```diff
@@ -2171,6 +2172,14 @@ def make_module_step(self, fake=False):
+            if module_only and os.path.exists(mod_filepath):
+               warning_msg = "Old module file found. Backing it up in %s. Diff is:\n%s"
+               mod_bck_filepath = back_up_file(mod_filepath, backup_extension="bck")
+               (mod_diff, _) = run_cmd("diff %s %s" % (mod_bck_filepath, mod_filepath))
```
use `diff -u`?

boegel commented on easybuild/framework/easyblock.py:
```diff
@@ -2171,6 +2172,14 @@ def make_module_step(self, fake=False):
+            module_only = build_option('module_only')
+            if module_only and os.path.exists(mod_filepath):
+               warning_msg = "Old module file found. Backing it up in %s. Diff is:\n%s"
+               mod_bck_filepath = back_up_file(mod_filepath, backup_extension="bck")
```
we should make this a hidden module, no?

also, this should be configurable, people should be able to disable the backup if they're confident enough (but I guess having it enabled by default makes sense)

damianam replied:
Making it a hidden module makes sense just for TCL modules. I had lua modules in mind, but since there are still people out there with TCL modules I agree with you.

Regarding making it configurable..... I am not so sure.

Pros:
1. Flexibility.

Cons:
1. Yet another option cluttering the `--help`
2. More code to maintain

However, if you are using TCL modules with Lmod, these will be recognized as modules that can be loaded, which is not what you want. Any other combination (TCL modules with Tmod, or lua modules with Lmod), will result in an invisible (to the modules tool) backup. So yes, I guess I should also prepare an option for it.....

boegel commented on easybuild/tools/filetools.py:
```diff
@@ -1168,16 +1198,12 @@ def move_logs(src_logfile, target_logfile):
-                _log.info("Moved existing log file %s to %s" % (new_log_path, oldlog_backup))
+                back_up_file(new_log_path)
 
+            # remove first to ensure portability
```
clarify portability?

boegel commented on easybuild/tools/filetools.py:
```diff
@@ -1150,6 +1151,35 @@ def rmtree2(path, n=3):
+
+
+def back_up_file(src_file, backup_extension=""):
+    """Backs up a file appending a backup extension and a number to it. Returns the name of the backup"""
```
clarify: the number is optional (only when existing backups are already there)

use `:param:` to document arguments

also support `hide` named argument to backup to a hidden file

boegel commented:
@damianam For functions you touch/change, please check whether they have a dedicated unit test already (and maybe enhance it while you're at it).

When adding new functions, add a dedicated test for them.

damianam commented:
@boegel I addressed all your comments. The code hasn't been tested yet, and the tests are still missing, but most of the bits and pieces should be there already.

damianam commented:
Tests are complete and code seems to be working. @easybuilders/easybuild-framework-maintainers  please review.

boegel commented on easybuild/framework/easyblock.py:
```diff
@@ -2202,8 +2202,30 @@ def make_module_step(self, fake=False):
+                else:
+                    self.log.info(warning_msg + ' No differences found.', mod_bck_filepath)
+                    print_warning(warning_msg % mod_bck_filepath + ' No differences found.')
+                self.log.info("Module file %s written: %s", mod_filepath, txt)
```
this whole block is rather big to just shove it into `make_module_step` like this, it's better to move it into a separate method named `back_up_existing_module_file` or something like that?

boegel commented on easybuild/framework/easyblock.py:
```diff
@@ -2202,8 +2202,30 @@ def make_module_step(self, fake=False):
+                warning_msg = "Old module file found. Backing it up in %s."
+                hidden = False
+                if isinstance(self.module_generator, ModuleGeneratorTcl):
+                     hidden = True
```
indent is off here?

Also, why only do hidden for Tcl modules?

The backup should always be a hidden module imho, you don't want to expose the backup module to users?

Maybe we should also back up outside of the module tree, e.g. in the current directory (or at a location specified via `--backup-modules`?

damianam replied:
Because lua modules have to end in `.lua`, otherwise they aren't treated as modules by Lmod, and don't show up with `ml av`. So there is no need to keep them hidden since the backup extension avoids that file ends in `.lua`.

Tcl modules on the other need to be hidden. Which is not that nice, since Lmod can still show them with `--show-hidden`. But that's the best compromise IMO. The alternative would be to remove the module header.

Backing up outside of the module tree kind of defeats the purpose of the backup. In my opinion the backup should be used for quick verification of the changes and for doing a "rollback" if necessary. If I have the current module and the backup in different files it slightly makes it more complicated and cumbersome, and more difficult to directly see if there was a previous version at any point with a simple `ls`

boegel commented on easybuild/framework/easyblock.py:
```diff
@@ -2202,8 +2202,30 @@ def make_module_step(self, fake=False):
+            module_only = build_option('module_only')
+            backup_modules = build_option('backup_modules')
+            if module_only and os.path.exists(mod_filepath) and backup_modules and not fake:
+                warning_msg = "Old module file found. Backing it up in %s."
```
I would move the `os.path.exists` check down into the `back_up_existing_module_file` method, and just exit early with a clear log message to say there's nothing to back up?

boegel commented on easybuild/framework/easyblock.py:
```diff
@@ -2202,8 +2202,30 @@ def make_module_step(self, fake=False):
+                if isinstance(self.module_generator, ModuleGeneratorTcl):
+                     hidden = True
+                else:
+                    hidden = False
```
`hidden` was set to `False` already...

boegel commented on easybuild/framework/easyblock.py:
```diff
@@ -2202,8 +2202,30 @@ def make_module_step(self, fake=False):
+                self.log.info("Module file %s written: %s", mod_filepath, txt)
+            else:
+                write_file(mod_filepath, txt)
+                self.log.info("Module file %s written: %s", mod_filepath, txt)
```
This is there twice now, why not just do this after backing up?

You would need delay the diff'ing a bit...

boegel commented on easybuild/framework/easyblock.py:
```diff
@@ -2202,8 +2202,30 @@ def make_module_step(self, fake=False):
+                (mod_diff, _) = run_cmd("diff -u %s %s" % (mod_bck_filepath, mod_filepath), log_ok=False, log_all=False)
+                if mod_diff:
+                    self.log.info(warning_msg + ' Diff is:\n%s', mod_bck_filepath, mod_diff)
+                    print_warning(warning_msg % mod_bck_filepath + ' Diff is:\n%s' % mod_diff)
```
The devil himself invented code duplication. Just construct the message once, and log/print afterwards:

```python
if mod_diff:
    diff_msg = (warning_msg + " Diff is:\n%s") % (mod_bck_filepath, mod_diff)
else:
    diff_msg = (warning_msg + "No differences found.") % mod_bck_filepath

self.log.info(diff_msg)
print_warning(diff_msg)

boegel commented on easybuild/tools/filetools.py:
```diff
@@ -1162,6 +1163,50 @@ def rmtree2(path, n=3):
         _log.info("Path %s successfully removed." % path)
 
+def find_backup_name_candidate(src_file):
+    """Returns a non-existing file to be used as destination for rotating backups"""
```
Not really rotating, I think, i.e. we don't recycle filenames after N backups?

boegel commented on easybuild/tools/filetools.py:
```diff
@@ -1162,6 +1163,50 @@ def rmtree2(path, n=3):
 
+def find_backup_name_candidate(src_file):
+    """Returns a non-existing file to be used as destination for rotating backups"""
+    dst_file = src_file
```
Hmm, shouldn't the default be `None`?

If the specified filepath doesn't exist, you can't back up that file either?

boegel commented on easybuild/tools/filetools.py:
```diff
@@ -1162,6 +1163,50 @@ def rmtree2(path, n=3):
+    if backup_extension:
+        trailing_string = ".%s" % backup_extension
+    else:
+        trailing_string = ""
```
Shorter/cleaner (imho):

```python
fn_pref, fn_suff = '', ''
if hidden:
    fn_pref = '.'
if backup_extension:
    fn_suff = '.%s' % backup_extension
```

Also, I prefer using `''` rather than `""` for strings without spaces (I find it easier on the eyes), but that's nitpicking ;)

boegel commented on easybuild/tools/filetools.py:
```diff
@@ -1162,6 +1163,50 @@ def rmtree2(path, n=3):
+    else:
+        trailing_string = ""
+
+    backup_file = find_backup_name_candidate("%s/%s%s%s" % \
```
```python
src_dir, src_fn = os.path.split(src_file)
backup_filepath = find_backup_name_candidate(os.path.join(src_dir, fn_pref + src_fn + fn_suff))

boegel commented on easybuild/tools/filetools.py:
```diff
@@ -1180,16 +1225,12 @@ def move_logs(src_logfile, target_logfile):
+                back_up_file(new_log_path)
 
+            # remove first to ensure portability (shutil.move might fail when overwriting files in some systems)
+            remove_file(new_log_path)
```
this should be part of `move_file` (under the condition that the target location already exists)?

boegel commented on easybuild/tools/module_generator.py:
```diff
@@ -185,7 +186,8 @@ def prepare(self, fake=False):
+        if os.path.exists(mod_filepath) and not build_option('extended_dry_run') and \
+                (not build_option('backup_modules') or fake):
             self.log.debug("Removing existing module file %s", mod_filepath)
             os.remove(mod_filepath)
```
Hmm, isn't this duplicate code since we're already removing the module file in `check_readiness_step` as well?

Also here: we should rename here already rather than not remove?

boegel commented on easybuild/tools/options.py:
```diff
@@ -324,6 +324,8 @@ def override_options(self):
             'allow-use-as-root-and-accept-consequences': ("Allow using of EasyBuild as root (NOT RECOMMENDED!)",
                                                           None, 'store_true', False),
+            'backup-modules': ("Back up an existing module file, if any. Only works when using --module-only",
+                            None, 'store_true', False),
```
align the `None` with the first `"` above

boegel commented on test/framework/filetools.py:
```diff
@@ -494,6 +494,53 @@ def test_guess_patch_level(self):
+    def test_back_up_file(self):
+        """Test back_up_file function."""
+        fh, fp = tempfile.mkstemp()
+        os.close(fh)
```
or just use `fp = os.path.join(self.test_prefix, 'test.txt')`

boegel commented on test/framework/filetools.py:
```diff
@@ -494,6 +494,53 @@ def test_guess_patch_level(self):
+
+        # Test simple file backup
+        ft.back_up_file(fp)
+        self.assertEqual(ft.read_file(fp+'_0'), 'foobar')
```
spaces around `+`

boegel commented on test/framework/filetools.py:
```diff
@@ -494,6 +494,53 @@ def test_guess_patch_level(self):
+        fh, fp = tempfile.mkstemp()
+        os.close(fh)
+
+        ft.write_file(fp, 'foobar')
```
define a variable for the file contents rather than copy-pasting `'foobar'` everywhere?

boegel commented on test/framework/filetools.py:
```diff
@@ -494,6 +494,53 @@ def test_guess_patch_level(self):
+        self.assertEqual(ft.read_file(fp), 'foobar')
+        # Test hidden simple file backup
+        ft.back_up_file(fp, hidden=True)
+        self.assertEqual(ft.read_file(os.path.dirname(fp)+'/.'+os.path.basename(fp)), 'foobar')
```
```python
dir, fn = os.path.split(fp)
self.assertEqual(ft.read_file(os.path.join(dir, '.' + fn)), test_txt)

boegel commented on test/framework/filetools.py:
```diff
@@ -494,6 +494,53 @@ def test_guess_patch_level(self):
+        ft.write_file(fp, 'foobar')
+
+        # Test simple file backup
+        ft.back_up_file(fp)
```
maybe check with `os.listdir` that you now only have two files, the original and the backup, nothing else?

boegel commented on test/framework/filetools.py:
```diff
@@ -494,6 +494,53 @@ def test_guess_patch_level(self):
+        self.assertEqual(ft.read_file(fp), 'foobar')
+        # Test simple file backup with extension
+        ft.back_up_file(fp, backup_extension="bck")
+        self.assertEqual(ft.read_file(fp+'.bck'), 'foobar')
```
spaces around `+`

boegel commented on test/framework/filetools.py:
```diff
@@ -494,6 +494,53 @@ def test_guess_patch_level(self):
+        self.assertEqual(ft.read_file(fp), 'foobar')
+        # Test hidden simple file backup with extension
+        ft.back_up_file(fp, backup_extension="bck", hidden=True)
+        self.assertEqual(ft.read_file(os.path.dirname(fp)+'/.'+os.path.basename(fp)+'.bck'), 'foobar')
```
use `os.path.split` & `os.path.join`, spaces around `+`

same remarks below

boegel commented on test/framework/toy_build.py:
```diff
@@ -1223,6 +1224,101 @@ def test_module_only(self):
+        toy_mod = os.path.join(self.test_installpath, 'modules', 'all', 'toy', '0.0-deps')
+
+        # only consider provided test modules
+        self.reset_modulepath([os.path.join(os.path.dirname(os.path.abspath(__file__)), 'modules')])
```
why is this needed?

boegel commented on test/framework/toy_build.py:
```diff
@@ -1223,6 +1224,101 @@ def test_module_only(self):
+        stderr = self.get_stderr()
+        self.mock_stderr(False)
+        self.assertTrue(os.path.exists(toy_mod))
+        self.assertTrue(os.path.exists(os.path.join(os.path.dirname(toy_mod), '.'+os.path.basename(toy_mod)+'.bck')))
```
use `os.path.split` & `os.path.join`, spaces around `+`

boegel commented on test/framework/toy_build.py:
```diff
@@ -1223,6 +1224,101 @@ def test_module_only(self):
+        self.eb_main(args + ['--backup-modules'], do_build=True, raise_error=True, verbose=True)
+        stderr = self.get_stderr()
+        self.mock_stderr(False)
+        self.assertTrue(os.path.exists(os.path.join(os.path.dirname(toy_mod), '.'+os.path.basename(toy_mod)+'.bck_0')))
```
use `os.path.split` & `os.path.join`, spaces around `+`

boegel commented on test/framework/toy_build.py:
```diff
@@ -1223,6 +1224,101 @@ def test_module_only(self):
+        warning = "WARNING: Old module file found. Backing it up in .* Diff is:"
+        diff = "some difference\n"
+        with open(toy_mod, "a") as fh:
+            fh.write(diff)
```
use `write_file` (it has an append mode)

boegel commented on test/framework/toy_build.py:
```diff
@@ -1223,6 +1224,101 @@ def test_module_only(self):
+        ]
+        args = common_args + ['--module-syntax=Tcl']
+
+        warning = "WARNING: Old module file found. Backing it up in .* No differences found"
```
Hmm, should this be a *warning*? It's just an informative message, no?

boegel commented on test/framework/toy_build.py:
```diff
@@ -1223,6 +1224,101 @@ def test_module_only(self):
+        if isinstance(self.modtool, Lmod):
+            args = common_args + ['--module-syntax=Lua']
+
+            toy_mod = os.path.join(self.test_installpath, 'modules', 'all', 'toy', '0.0-deps.lua')
```
rather than copy-pasting all the checks, we should hoist them in a helper function, and simplify the test to

```python
def test_backup_modules(self):
    """Test use of --backup-modules."""
    def run_checks():
        ...

    ...

    toy_mod_fn = '0.0-deps'
    run_checks()

    if isinstance(self.modtool, Lmod):
      toy_mod_fn = '0.0-deps.lua'
      run_checks()

boegel commented:
Going in, thanks @damianam!
